Hint spawn fix

rcorre opened the pull request.

rcorre commented:
I wanted to test `--rapid` as well, but this test ends up in a recursive loop:

    Scenario: Using :hint spawn with flags (issue 797)
        When I open data/hints/link.html
        And I run :hint --rapid all spawn -v echo
        And I run :follow-hint a
        Then the message "Command exited successfully." should be shown

```
...
IGNORED:   File "/data/project/contrib/qutebrowser/qutebrowser/browser/hints.py", line 949, in fire
IGNORED:     self.filter_hints(None)
IGNORED:   File "/data/project/contrib/qutebrowser/qutebrowser/browser/hints.py", line 892, in filter_hints
IGNORED:     self.fire(*visible)
IGNORED:   File "/data/project/contrib/qutebrowser/qutebrowser/browser/hints.py", line 949, in fire
IGNORED:     self.filter_hints(None)
IGNORED:   File "/data/project/contrib/qutebrowser/qutebrowser/browser/hints.py", line 892, in filter_hints
IGNORED:     self.fire(*visible)
IGNORED:   File "/data/project/contrib/qutebrowser/qutebrowser/browser/hints.py", line 936, in fire
IGNORED:     url = self._resolve_url(elem, self._context.baseurl)
IGNORED:   File "/data/project/contrib/qutebrowser/qutebrowser/browser/hints.py", line 599, in _resolve_url
IGNORED:     text = elem[attr]
IGNORED:   File "/data/project/contrib/qutebrowser/qutebrowser/browser/webelem.py", line 134, in __getitem__
IGNORED:     if key not in self:
INVALID: RecursionError: maximum recursion depth exceeded while calling a Python object
IGNORED:
```

Weirdly, replacing the `Then` line with `Then no crash should happen` is fine...

The-Compiler commented:
I can reproduce the `RecursionError`, I'll look into it.

The reason that `No crash should happen` worked is because it was "too fast", and terminated the subprocess before it had a chance to crash (because the tests finished). I added a small delay in 
107934e4e1468b98b4a9849a4dd6949a03fcc2fd.

The-Compiler commented:
I can also reproduce this without your PR, I opened #1513.

The-Compiler commented:
Fixed now too - can you rebase on master and add that test please?

The-Compiler commented on qutebrowser/browser/hints.py:
```diff
@@ -581,9 +580,8 @@ def _spawn(self, url, context):
-        proc = guiprocess.GUIProcess(self._win_id, what='command', parent=self)
-        proc.start(cmd, args)
+        commandrunner = runners.CommandRunner(self._win_id)
+        commandrunner.run_safely('spawn ' + ' '.join(args))
```
Clever solution! I think it'd need to call [`shlex.quote`](https://docs.python.org/dev/library/shlex.html#shlex.quote) on the arguments though, so it works fine with a commandline containing spaces? I'm not 100% sure though with all the different splitting involved, would be nice to have a test added.

The-Compiler commented on qutebrowser/browser/hints.py:
```diff
@@ -737,7 +735,7 @@ def follow_prevnext(self, frame, baseurl, prev=False, tab=False,
 
     @cmdutils.register(instance='hintmanager', scope='tab', name='hint',
-                       star_args_optional=True)
+                       star_args_optional=True, maxsplit=2)
```
How does this affect the `userscript`/`run`/`fill` targets?

The-Compiler commented:
Note I also merged #1311 now - that means `hints/link.html` got renamed to `hints/html/simple.html`.

rcorre commented:
Rebased and added tests. There is still no test for `:hint <...> userscript ...` as this does not get redirected to `tests/integration/data/userscripts` like 'I execute the userscript ...` does (it goes to the 'real' userscript directory)

The-Compiler commented:
You should be able to pass an absolute path though - probably time to add a `(testdir)` substitution or so, similar to `(port)`. I'll take a look at it tomorrow.

The Windows/AppVeyor build seemed to fail for an unrelated reason, I triggered a rebuild now.

The-Compiler commented:
Thanks!
